Parser

ginabolo opened the pull request.

Commit:
Implement parser

Commit:
Use specific error classes

Commit:
Carry through "some changes" by Brian, fix compilation issues

Commit:
Static methods are overrated

Commit:
Tokenizer also throws ParseErrors

kizggerg commented:
When you run `npm test`, do all of the parser tests pass?

ginabolo commented:
Seems like it crashes due to a circular dependency between SymbolTable and FlowSymbolTable? I'm getting
`export default class FlowSymbolTable extends SymbolTable {`
`TypeError: Class extends value undefined is not a function or null`

kizggerg commented on src/symbol_table/SymbolTable.ts:
```diff
@@ -2,6 +2,7 @@ import {Action, Flow} from "../ast";
 import Value from "../ast/Value";
 import FlowSymbolTable from "./FlowSymbolTable";
```
@ginabolo Try removing this import. I think this is where the circular dependency is coming in.

ginabolo replied:
Yes but don't I need that so that the flows can map to FlowSymbolTables? The SymbolTable imports FlowSymbolTable, and FlowSymbolTable obviously imports the SymbolTable because it extends it.

ginabolo replied:
Okay I moved both definitions to the same file ¯\_(ツ)_/¯
Working on passing the tests.

kizggerg replied:
This is fine for now, but not best practice.
In general, you shouldn't need to import FlowSymbolTable in SymbolTable.
This problem is typically solved through interface segregation, or using generic types: https://www.typescriptlang.org/docs/handbook/generics.html

kizggerg replied:
( But don't worry about fixing the above for now - our top priority is getting a working parser out the door ;) )

Commit:
Fix many parsing bugs

Commit:
Parsing tests except imports pass

Commit:
Move symbol table populating to typechecker to disallow referencing objects before they are declared

Commit:
Fix typechecking bugs

Limitations:
- Imports not implemented
- (Temporarily?) Allow modifies clauses before their variable declarations
- (Temporarily?) Allow a flow, an action and a primitive variable to share an identifier name

kizggerg commented:
👍 LGTM
